library/Faktory/JobOptions.hs

```diff
@@ -0,0 +1,85 @@
+{-# LANGUAGE DerivingVia #-}
+
+module Faktory.JobOptions
+  ( JobOptions(..)
+
+  -- * Modifiers
+  , retry
+  , once
+  , queue
+  , jobtype
+  , at
+  , in_
+
+  -- * Enqueue-time modifiers
+  , setAtFromIn
+  , namespaceQueue
+  ) where
+
+import Faktory.Prelude
+
+import Data.Aeson
+import Data.Aeson.Casing
+import Data.Semigroup (Last(..))
+import Data.Semigroup.Generic
+import Data.Time
+import Faktory.Settings (Namespace, Queue)
+import qualified Faktory.Settings as Settings
+import GHC.Generics
+
+-- | Options for the execution of a job
+--
+-- These can be constructed using '(<>)':
+--
+-- @
+-- let options = 'retry' 1 <> 'jobtype' "MyJob"
+-- @
+--
+-- To enqueue with defaults, use 'mempty'.
+--
+-- Options use 'Last' semantics, so use @modifier <>@ to set something only if
+-- not already set.
+--
+data JobOptions = JobOptions
+  { joJobtype :: Maybe (Last String)
+  , joRetry :: Maybe (Last Int)
+  , joQueue :: Maybe (Last Queue)
+  , joAt :: Maybe (Last UTCTime)
+  , joIn :: Maybe (Last NominalDiffTime)
+  }
+  deriving stock Generic
+  deriving (Semigroup, Monoid) via GenericSemigroupMonoid JobOptions
+
+instance FromJSON JobOptions where
+  parseJSON = genericParseJSON $ aesonPrefix snakeCase
+
+setAtFromIn :: JobOptions -> IO JobOptions
+setAtFromIn options = case joIn options of
+  Nothing -> pure options
+  Just (Last nd) -> do
+    time <- addUTCTime nd <$> getCurrentTime
+    pure $ at time <> options
```

Comment:
should this be `options <> at time`?

Reply:
It's arbitrary, but I chose for `in_ x <> at y` (or `at y <> in_ x`) to always respect the `at` and never override it from `in_`. Therefore, this needs to be `at time <> options`. There's a test of this behavior too.

Reply:
We could make to so that "Last semantics" actually applies to at-vs-in as well, by making

```hs
at t = mempty
  { joAt = Just (Last t)
  , joIn = Nothing
  }

in_ d = mempty
  { joAt = Nothing
  , joIn = Just (Last d)
  }
```

Then the choice here won't matter.

WDYT?

Reply:
I think whatever semantics we pick should be consistent, which would mean not having to remember special interactions between fields. I think that `Nothing` you're doing there is the default anyway, so that wouldn't have `Last` semantics anyway. Maybe you can make the options have one field with a sum type instead of two fields?

```haskell
at t = mempty
  { joSchedule = Just $ Last $ Left t
  }

in_ d = mempty
  { joSchedule = Just $ Last $ Right d
  }
```

And then in `applyOptions` you still handle the `Right` case as before

Reply:
Yeah, that does what I would consider the right thing:

```haskell
data Thing = Thing
  { x :: Maybe (Last Int)
  , y :: Maybe (Last (Either Int Bool))
  }
  deriving (Eq, Show, Generic)

instance Semigroup Thing where
  (<>) = gmappend

instance Monoid Thing where
  mempty = gmempty

xInt :: Int -> Thing
xInt n = mempty { x = Just $ Last n }

yInt :: Int -> Thing
yInt n = mempty { y = Just $ Last $ Left n }

yBool :: Bool -> Thing
yBool b = mempty { y = Just $ Last $ Right b }

-- > xInt 3
-- Thing {x = Just (Last {getLast = 3}), y = Nothing}
-- > xInt 3 <> yInt 4
-- Thing {x = Just (Last {getLast = 3}), y = Just (Last {getLast = Left 4})}
-- > xInt 3 <> yInt 4 <> yInt 12
-- Thing {x = Just (Last {getLast = 3}), y = Just (Last {getLast = Left 12})}
-- > xInt 3 <> yInt 4 <> yInt 12 <> yBool True
-- Thing {x = Just (Last {getLast = 3}), y = Just (Last {getLast = Right True})}
-- > xInt 3 <> yInt 4 <> yInt 12 <> yBool True <> yBool False
-- Thing {x = Just (Last {getLast = 3}), y = Just (Last {getLast = Right False})}
```

Reply:
> I think whatever semantics we pick should be consistent, which would mean not having to remember special interactions between fields

Regardless of implementation don't you always need to remember interactions between `at` and `in_`?

The interface is always the same,

```
at x <> in_ y

in_ y <> at x
```

Will interact in some way that you have to remember:

- PR as-is: always the same behavior (`at` always wins), which is "consistent" on one axis
- My `Nothing` suggestion: last wins, which is consistent on another axis (behavior like using the same option twice)
- Your suggestion: also last wins, just a different internal representation

Reply:
I like "last always" wins because it applies to every option, not just the two that are related. Again, I don't think your `Nothing` solution actually works because `Just (Last x) <> Nothing` -> `Just (Last x)`

Reply:
To clarify, I think my `Nothing` suggestion is a `(Maybe a, Maybe b)` type where it's guaranteed to be `(Nothing, Nothing)`, `(Just a, Nothing)`, or `(Nothing, Just b)` by construction. So, isomorphic to your `Maybe (Either a b)`

```
Nothing ~ Nothing
(Just a, Nothing) ~ Just (Left a)
(Nothing, Just b) ~ Just (Right b)
```

Through this lens, the `Either` representation definitely makes more sense. So, if we do want those semantics over the PR as-is, I'll use that implementation.

Reply:
> Again, I don't think your Nothing solution actually works because

Ah, I see that now. Well, then there you go.